addons/io_hubs_addon/components/definitions/morph_audio_feedback.py

```diff
@@ -0,0 +1,74 @@
+import atexit
+import bpy
+from bpy.props import FloatProperty, StringProperty, EnumProperty
+from ..hubs_component import HubsComponent
+from ..types import Category, PanelType, NodeType
+
+
+def get_shape_keys(self, context):
+    shape_keys = []
+    if context.object.data.shape_keys:
+        for item in context.object.data.shape_keys.key_blocks:
+            shape_keys.append((item.name, item.name, ""))
+    return shape_keys
+
+
+class MorphAudioFeedback(HubsComponent):
+    _definition = {
+        'name': 'morph-audio-feedback',
+        'display_name': 'Morph Audio Feedback',
+        'category': Category.AVATAR,
+        'node_type': NodeType.NODE,
+        'panel_type': [PanelType.OBJECT],
+        'icon': 'MOD_SMOOTH'
+    }
+
+    name: StringProperty(
+        name="Name",
+        description="Name"
+    )
+
+    shape_key: EnumProperty(
+        name="Shape Key",
+        description="Shape key to morph",
+        items=get_shape_keys
+    )
+
+    minValue: FloatProperty(name="Min Value",
+                            description="Min Value",
+                            default=0.0,)
+
+    maxValue: FloatProperty(name="Max Value",
+                            description="Max Value",
+                            default=1.0)
+
+    @classmethod
+    def poll(cls, context):
+        return context.object.type == 'MESH'
+
+    @classmethod
+    def migrate(cls, version):
+        if version < (0, 1, 0):
+            for ob in bpy.data.objects:
+                if cls.get_name() in ob.hubs_component_list.items:
+                    ob.hubs_component_morph_audio_feedback.shape_key = (
+                        ob.hubs_component_morph_audio_feedback.name)
+
+    def draw(self, context, layout):
+        shape_keys = context.object.data.shape_keys
+        if shape_keys and len(shape_keys.key_blocks) > 0:
+            layout.prop(data=self, property="shape_key")
+            layout.prop(data=self, property="minValue")
+            layout.prop(data=self, property="maxValue")
+        else:
+            col = layout.column()
+            col.alert = True
+            col.label(text='No shape keys available',
+                      icon='ERROR')
+
+    def gather(self, export_settings, object):
+        return {
+            'name': object.hubs_component_morph_audio_feedback.shape_key,
+            'minValue': object.hubs_component_morph_audio_feedback.minValue,
+            'maxValue': object.hubs_component_morph_audio_feedback.maxValue
+        }
```

Comment:
A couple things:
- Enums require that "Python must keep a reference to the strings returned by the [items] callback or Blender will misbehave or even crash.".  (You can see how to work around this here: https://developer.blender.org/T50426#412032)
- If the previous morph-audio-feedback component value does not match any of the shape keys it will cause an error and won't show anything in the list.
- The Basis shape key isn't valid for this use when shape keys are relative (if they're not relative then they're played one after another in sequence, which I don't think will work with this component?).  Maybe check with Dom and Jim, etc. to make sure, but I'm pretty sure the Basis key should be excluded.

I've addressed all these points in the code below, as well as added in a check for commas and spaces in the shape key names that I'm pretty sure would cause export issues (loop-animation should have this warning as well, but it's not critical and can happen after the beta.  I only added it here because I was changing everything else).

So see what you think, and feel free to modify/do something else if you want.

```suggestion
import atexit
import bpy
from bpy.props import FloatProperty, StringProperty, EnumProperty
from ..hubs_component import HubsComponent
from ..types import Category, PanelType, NodeType

shape_keys = []

def get_shape_keys(self, context):
    global shape_keys
    shape_keys = []
    count = 0

    if self.name:
        shape_keys.append((f"ERROR: {self.name}", f"ERROR: {self.name}", "No matching shape key found", 'ERROR', count))
        count += 1

    if context.object.data.shape_keys:
        for item in context.object.data.shape_keys.key_blocks:
            if ' ' in item.name or ',' in item.name:
                shape_keys.append((f"ERROR: {item.name}", f"ERROR: {item.name}", "Shape keys can't have spaces or commas in them", 'ERROR', count))
                count += 1
            elif item == item.relative_key:
                pass
            else:
                shape_keys.append((item.name, item.name, "", 'SHAPEKEY_DATA', count))
                count += 1

    return shape_keys

def set_shape_key(self, value):
    global shape_keys

    self["shape_key"] = value

    if self.name and f"ERROR_{self.name}" != shape_keys[value][0]:
        self.name = ''


class MorphAudioFeedback(HubsComponent):
    _definition = {
        'name': 'morph-audio-feedback',
        'display_name': 'Morph Audio Feedback',
        'category': Category.AVATAR,
        'node_type': NodeType.NODE,
        'panel_type': [PanelType.OBJECT],
        'icon': 'MOD_SMOOTH'
    }

    name: StringProperty(
        name="Name",
        description="Name"
    )

    shape_key: EnumProperty(
        name="Shape Key",
        description="Shape key to morph",
        items=get_shape_keys,
        set=set_shape_key
    )

    minValue: FloatProperty(name="Min Value",
                            description="Min Value",
                            default=0.0,)

    maxValue: FloatProperty(name="Max Value",
                            description="Max Value",
                            default=1.0)

    @classmethod
    def poll(cls, context):
        return context.object.type == 'MESH'

    @classmethod
    def migrate(cls, version):
        if version < (0, 1, 0):
            for ob in bpy.data.objects:
                if cls.get_name() in ob.hubs_component_list.items:
                    component = ob.hubs_component_morph_audio_feedback
                    if component.name in get_shape_keys(component, bpy.context):
                        component.shape_key = (component.name)
                        component.name = ''
                    else:
                        component.shape_key = (f"ERROR_{component.name}")

    def draw(self, context, layout):
        if get_shape_keys(self, context):
            layout.prop(data=self, property="shape_key")
            layout.prop(data=self, property="minValue")
            layout.prop(data=self, property="maxValue")
        else:
            col = layout.column()
            col.alert = True
            col.label(text='No shape keys available',
                      icon='ERROR')

    def gather(self, export_settings, object):
        return {
            'name': object.hubs_component_morph_audio_feedback.shape_key,
            'minValue': object.hubs_component_morph_audio_feedback.minValue,
            'maxValue': object.hubs_component_morph_audio_feedback.maxValue
        }
```

Reply:
Thanks for that. I've made some updates to this so I've landed a different commit. I've added a specific error label and removed the ERROR string from the item text, I feel that's it's a bit clearer this way.
Also it seems that if you override the set callback in the `EnumProperty` you also need to override the get to return the stored value.

Reply:
You're welcome.  Yeah, your UI is very nice.
Unfortunately, I've discovered a number of issues with the EnumProperty, that I didn't anticipate and aren't easily worked around.

- morph-audio-feedback doesn't pass along the shape key to export if it hasn't been interacted with since migration (or maybe program start, I'm not sure) - seems to be a problem with the get function, and it appears to be on the blender side.
- the morph-audio-feedback code errors out if there are no shape keys in some cases.
- there's an error in migration because it uses the active object, not the object with the component.  (I have a file where the avatar armature is the active object and it doesn't like that at all when migrating)
- deleting shape keys can cause which one is used in the component to change (this is because it's actually storing indexes, rather than names)

As I was attempting to fix the above issues, I ran into others caused by my partial fixes, but I've finally gotten it down to two issues left (hopefully), and I added in a blank item so it'll fall back to something predictable if you delete the selected shape key (plus you aren't forced to use a key if you don't want to while developing your avatar).
The known issues in my code:
- the shape key will be lost if it's renamed.
- blender errors on Object rna removed if you make shape keys on the default cube and set one in morph-audio-feedback, then open import_dev_test_8 ([import_dev_test_8.zip](https://github.com/MozillaReality/hubs-blender-exporter/files/8908913/import_dev_test_8.zip))

And here is the current contents of my morph-audio-feedback.py:
```
import atexit
import bpy
from bpy.props import IntProperty, FloatProperty, StringProperty, EnumProperty, PointerProperty
from ..hubs_component import HubsComponent
from ..types import Category, PanelType, NodeType

BLANK_SHAPE_KEY_IDENTIFIER = '56489845649846498select_a_shape_key8498416549846849849843'
shape_keys = []


def get_shape_keys(self, context):
    global BLANK_SHAPE_KEY_IDENTIFIER
    global shape_keys
    shape_keys = []
    ob_shape_keys = self.ob.data.shape_keys
    count = 0

    shape_keys.append((BLANK_SHAPE_KEY_IDENTIFIER, 'Select a Shape Key',
                                  "Select a Shape Key", 'BLANK', count))

    count += 1

    if (self.name and not ob_shape_keys) or ((self.name and ob_shape_keys) and self.name not in ob_shape_keys.key_blocks):
        shape_keys.append((self.name, self.name,
                          "No matching shape key found", 'ERROR', count))
        count += 1
    if ob_shape_keys:
        for item in ob_shape_keys.key_blocks:
            if ' ' in item.name or ',' in item.name:
                shape_keys.append((item.name, item.name,
                                  "Shape key names can't have spaces or commas", 'ERROR', count))
                count += 1
            elif item == item.relative_key:
                pass
            else:
                shape_keys.append(
                    (item.name, item.name, "", 'SHAPEKEY_DATA', count))
                count += 1

    return shape_keys


def get_shape_key(self):
    global BLANK_SHAPE_KEY_IDENTIFIER
    global shape_keys

    #if self.active_shape_key_index > len(shape_keys) - 1:
        #self.active_shape_key_index = 0
        #self.active_shape_key_name = BLANK_SHAPE_KEY_IDENTIFIER
    if self.active_shape_key_index > (len(shape_keys) -1) or shape_keys[self.active_shape_key_index][0] != self.active_shape_key_name:
        shape_key_names = [item[0] for item in shape_keys]
        if self.active_shape_key_name in shape_key_names:
            self.active_shape_key_index = shape_key_names.index(self.active_shape_key_name)
        else:
            self.active_shape_key_index = 0
            self.active_shape_key_name = BLANK_SHAPE_KEY_IDENTIFIER



    return self.active_shape_key_index


def set_shape_key(self, value):
    global shape_keys
    if self.name and self.name != shape_keys[value][0] and value != 0:
        self.name = ''
        self.active_shape_key_index = value - 1
        self.active_shape_key_name = shape_keys[value - 1][0]
    else:
        self.active_shape_key_index = value
        self.active_shape_key_name = shape_keys[value][0]


class MorphAudioFeedback(HubsComponent):
    _definition = {
        'name': 'morph-audio-feedback',
        'display_name': 'Morph Audio Feedback',
        'category': Category.AVATAR,
        'node_type': NodeType.NODE,
        'panel_type': [PanelType.OBJECT],
        'icon': 'MOD_SMOOTH'
    }

    name: StringProperty(
        name="Name",
        description="Name"
    )

    active_shape_key_index: IntProperty(default=0, options={'HIDDEN'})

    active_shape_key_name: StringProperty(default='', options={'HIDDEN'})

    shape_key: EnumProperty(
        name="Shape Key",
        description="Shape key to morph",
        items=get_shape_keys,
        get=get_shape_key,
        set=set_shape_key
    )

    minValue: FloatProperty(name="Min Value",
                            description="Min Value",
                            default=0.0,)

    maxValue: FloatProperty(name="Max Value",
                            description="Max Value",
                            default=1.0)

    ob: PointerProperty(type=bpy.types.Object, options={'HIDDEN'})

    @classmethod
    def init(cls, obj):
        cls.ob = obj

    @classmethod
    def poll(cls, context):
        return context.object.type == 'MESH'

    @classmethod
    def migrate(cls, version):
        if version < (1, 0, 0):
            for ob in bpy.data.objects:
                if cls.get_name() in ob.hubs_component_list.items:
                    component = ob.hubs_component_morph_audio_feedback
                    component.ob = ob
                    if component.name in [item[0] for item in get_shape_keys(component, bpy.context) if item[3] != 'ERROR']:
                        component.shape_key = (component.name)
                        component.name = ''
                    else:
                        component.shape_key = (component.name)

    def draw(self, context, layout):
        global BLANK_SHAPE_KEY_IDENTIFIER

        if len(get_shape_keys(self, context)) > 1:
            layout.prop(data=self, property="shape_key")
            ob_shape_keys = context.object.data.shape_keys
            if (self.shape_key != BLANK_SHAPE_KEY_IDENTIFIER) and ((self.shape_key and not ob_shape_keys) or (ob_shape_keys and self.shape_key not in ob_shape_keys.key_blocks)):
                col = layout.column()
                col.alert = True
                col.label(text="No matching shape key found",
                          icon='ERROR')
            elif ' ' in self.shape_key or ',' in self.shape_key:
                col = layout.column()
                col.alert = True
                col.label(text="Shape key names can't have spaces or commas",
                          icon='ERROR')
            layout.prop(data=self, property="minValue")
            layout.prop(data=self, property="maxValue")
        else:
            col = layout.column()
            col.alert = True
            col.label(text='No shape keys available',
                      icon='ERROR')

    def gather(self, export_settings, object):
        global BLANK_SHAPE_KEY_IDENTIFIER

        component = object.hubs_component_morph_audio_feedback
        name = component.shape_key if component.shape_key != BLANK_SHAPE_KEY_IDENTIFIER else ''
        minValue = component.minValue
        maxValue = component.maxValue
        return {
            'name': name,
            'minValue': minValue,
            'maxValue': maxValue
        }
```